### PR TITLE
[Feature][Connector-V2] Support user-defined schema for source connectors

### PR DESCRIPTION
**Pull request by TyrantLucifer**
## Purpose of this pull request

Add a common function for source connectors, now source connectors can use this function to parse user-defined schema information in their own config.

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
@Hisoka-X @ashulin, please review, thx.

File: seatunnel-connectors-v2/connector-common/pom.xml [hunk 1]
```diff
@@ -38,5 +38,10 @@
             <artifactId>seatunnel-api</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <scope>test</scope>
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-common/pom.xml]: redundant statement

Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/pom.xml]: Fixed.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+import org.apache.seatunnel.shade.com.typesafe.config.Config;
+import org.apache.seatunnel.shade.com.typesafe.config.ConfigValue;
+
+import java.util.Map;
+
+public class SeatunnelSchema {
+    private static final String FIELD_KEY = "fields";
+    private static final String FORMAT_KEY = "format";
+    private static final String DELIMITER_KEY = "delimiter";
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: This field not a common field, it only work from text or csv so on. Maybe you should use inheritance to split different types

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+        if (type.contains("<") || type.contains(">")) {
+            // Map type or Array type
+            int start = type.indexOf("<");
+            int end = type.lastIndexOf(">");
+            genericType = type
+                    // get the content between '<' and '>'
+                    .substring(start + 1, end)
+                    // replace the space between key and value
+                    .replace(" ", "");
+            if (type.contains("MAP")) {
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]:
```suggestion
            if (type.startsWith(SqlType.MAP.name());) {
```
Array & Decimal type in the same way.

**Review by ashulin — COMMENTED** [review]
Now the code is more complex. Can you split `Decimal`, `Array`, `Map` in `SeatunnelSchema#parseTypeByString` into their own methods?

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+import org.apache.seatunnel.common.config.CheckResult;
+
+import org.apache.seatunnel.shade.com.typesafe.config.Config;
+import org.apache.seatunnel.shade.com.typesafe.config.ConfigValue;
+
+import java.util.Map;
+
+public class SeatunnelSchema {
+    private static final String FIELD_KEY = "fields";
+    private static final String FORMAT_KEY = "format";
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: I think the schema only expresses the structure of the table, and the format should be defined and supported by the connector.

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: Use `seatunnel-formats`?

Comment (ashulin) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: For mysql and oracle connectors, users can also want to define schema, in which case format is invalid.
 I think format should be another feature (in fact, there is currently `json format` in the `seatunnel-format-json` module);

Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: So your guys think format information should be taken out in this class and abstract format information out to another feature?

Comment (ashulin) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: The format feature already exists, but it is not yet enabled on the entire process.

Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: OK, I will only keep schema information in it.

Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: @ashulin @Hisoka-X @CalvinKirs pls review thx.

File: seatunnel-connectors-v2/connector-common/src/test/java/org/apache/seatunnel/connector/common/schema/SchemaParseTest.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    @Test
+    public void testSimpleSchemaParse() throws FileNotFoundException, URISyntaxException {
+        String path = getTestConfigFile("/simple.schema.conf");
+        Config config = ConfigFactory
+                .parseFile(new File(path))
+                .resolve(ConfigResolveOptions.defaults().setAllowUnresolved(true))
+                .resolveWith(ConfigFactory.systemProperties(), ConfigResolveOptions.defaults().setAllowUnresolved(true));
+        config = config.getConfig("schema");
+        SeatunnelSchema seatunnelSchema = SeatunnelSchema.buildWithConfig(config);
+        Assertions.assertNotNull(seatunnelSchema);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-common/src/test/java/org/apache/seatunnel/connector/common/schema/SchemaParseTest.java:52]: The assert method too simple. You should make sure the seatunnel schema content is match with schema config. Pick some field and assert it.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java (hunk 4 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: By the way, now we didn't contain formation information in our config, so whether config should be changed as the following:

```hcon
schema {
    map = "map<string, string>"
    array = "array<tinyint>"
    string = string
    boolean = boolean
    tinyint = tinyint
    smallint = smallint
    int = int
    bigint = bigint
    float = float
    double = double
    decimal = "decimal(30, 8)"
    null = "null"
    bytes = bytes
    date = date
    time = time
    timestamp = timestamp
}
```

Waiting for your guys advice.

File: seatunnel-connectors-v2/connector-common/src/test/java/org/apache/seatunnel/connector/common/schema/SchemaParseTest.java (hunk 5 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/test/java/org/apache/seatunnel/connector/common/schema/SchemaParseTest.java:52]: Resolved. Please review, thx.

File: seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java (hunk 3 above)
Comment (TyrantLucifer) [reply seatunnel-connectors-v2/connector-common/src/main/java/org/apache/seatunnel/connectors/seatunnel/common/schema/SeatunnelSchema.java]: Resolved, please review, thx.

**Review by Hisoka-X — APPROVED** [review]
LGTM

**Review by ashulin — COMMENTED** [review]
LGTM. Thank you for your contribution to Apache SeaTunnel!

**Pull request merged**